File: libtmux/src/main/java/io/github/libtmux/CommandChain.java
package io.github.libtmux;

import io.github.libtmux.batch.Batch;
import io.github.libtmux.batch.BatchResult;
import java.util.List;

/**
 * A sequence of tmux commands where each one acts on what the last one made.
 *
 * <p>tmux moves its own current target as a group runs: a {@code new-window} makes the new window
 * current, a following {@code split-window} splits that window, and a following {@code send-keys}
 * types into the pane that split produced. None of these steps names a target, and that is the
 * point — the alternative is a round trip per step to learn the id of the thing just created.
 *
 * <p>The whole chain is one tmux invocation, and every step is attributed individually, so a chain
 * that fails halfway says which step failed and which never ran.
 */
public final class CommandChain {

    private final Batch batch;

    CommandChain(Batch batch) {
        this.batch = batch;
    }

    /** Creates a window and makes it the one following steps act on. */
    public CommandChain newWindow(String name) {
        return then("new-window", "-n", name);
    }

    /** Renames the current window. */
    public CommandChain renameWindow(String name) {
        return then("rename-window", name);
    }

    /**
     * Splits the current pane into a left and a right one.
     *
     * <p>Named for what it produces rather than for tmux's {@code -h}, which reads as though it
     * described the divider.
     */
    public CommandChain splitLeftRight() {
        return then("split-window", "-h");
    }

    /** Splits the current pane into a top and a bottom one. */
    public CommandChain splitTopBottom() {
        return then("split-window", "-v");
    }

    /** Types a line into the current pane and presses Enter, which is how a command gets run. */
    public CommandChain sendLine(String command) {
        return then("send-keys", command, "Enter");
    }

    /**
     * Arranges the current window.
     *
     * @throws IllegalArgumentException if tmux would not recognise the layout, which on some
     *     versions ends the whole server rather than the command
     */
    public CommandChain arrange(String layout) {
        return then("select-layout", Layouts.require(layout));
    }

    /** Adds any tmux command, for whatever this class does not name. */
    public CommandChain then(String... argv) {
        batch.add(argv);
        return this;
    }

    /** Adds any tmux command. */
    public CommandChain then(List<String> argv) {
        batch.add(argv);
        return this;
    }

    /** Runs the whole chain in one tmux invocation, attributing each step. */
    public BatchResult run() {
        return batch.run();
    }
}

File: scala/cats/src/main/scala/io/github/libtmux/scaladsl/cats/CommandChain.scala
package io.github.libtmux.scaladsl.cats

import io.github.libtmux.scaladsl.{BatchResult, blocking}

/** An immutable plan using tmux's evolving current target. Construction does no
  * tmux I/O; each run builds and executes a fresh Java chain.
  */
final class CommandChain[F[_]] private[cats] (
    create: () => blocking.CommandChain,
    execution: Execution[F],
    steps: Vector[blocking.CommandChain => blocking.CommandChain] = Vector.empty
) {
  private def append(
      step: blocking.CommandChain => blocking.CommandChain
  ): CommandChain[F] = new CommandChain(create, execution, steps :+ step)

  def newWindow(name: String): CommandChain[F] = append(_.newWindow(name))
  def renameWindow(name: String): CommandChain[F] = append(_.renameWindow(name))
  def splitLeftRight(): CommandChain[F] = append(_.splitLeftRight())
  def splitTopBottom(): CommandChain[F] = append(_.splitTopBottom())
  def sendLine(command: String): CommandChain[F] = append(_.sendLine(command))

  /** Layout validation and its possible version query occur during run. */
  def arrange(layout: String): CommandChain[F] = append(_.arrange(layout))

  def andThen(argv: Seq[String]): CommandChain[F] = {
    val copied = argv.toVector
    append(_.andThen(copied))
  }

  def andThen(command: String, arguments: String*): CommandChain[F] =
    andThen(command +: arguments)

  def run: F[BatchResult] = execution {
    steps.foldLeft(create())((chain, step) => step(chain)).run()
  }
}
